src/robothub_oak/application.py

```diff
@@ -1,162 +1,257 @@
-import logging as log
+import atexit
+import logging
+import signal
+import threading
 import time
+
 from abc import ABC, abstractmethod
 from threading import Thread
-from typing import Dict, Optional
+from typing import Optional
 
 import robothub_core
 from depthai_sdk import OakCamera
 
 from robothub_oak.utils import get_device_performance_metrics, get_device_details
 
-__all__ = ['BaseApplication']
+__all__ = ["BaseApplication"]
+
+logger = logging.getLogger(__name__)
 
 
 class BaseApplication(robothub_core.RobotHubApplication, ABC):
+    """
+    This class acts as the main entry point for the user, managing a single device, creating pipelines,
+    and polling the device for new data. Derived classes must implement the `setup_pipeline` method.
+
+    Attributes:
+        config: The configuration settings from the robotapp.toml config file.
+
+    Methods:
+        setup_pipeline: Abstract method to be implemented by child classes. Sets up the pipeline for a device.
+        on_device_connected: Optional method, called when a device is connected.
+        on_device_disconnected: Optional method, called when a device is disconnected.
+        on_stop: Optional method, called when the application is stopped.
+    """
+
     def __init__(self):
         robothub_core.RobotHubApplication.__init__(self)
         ABC.__init__(self)
 
-        self.__polling_threads = []
-        self.__reporting_threads = []
-        self.__camera_threads = []
-        self.__devices: Dict[str, Optional[OakCamera]] = {}
-        self.__device_states: Dict[str, robothub_core.DeviceState] = {}
-
         self.config = robothub_core.CONFIGURATION
 
-    def on_start(self) -> None:
-        for device in robothub_core.DEVICES:
-            device_mxid = device.oak['serialNumber']
-            self.__devices[device_mxid] = None
-            self.__device_states[device_mxid] = robothub_core.DeviceState.DISCONNECTED
+        self.__rh_device: Optional[robothub_core.RobotHubDevice] = None
+        self.__device_mxid: Optional[str] = None
+        self.__device_product_name: Optional[str] = None
+        self.__device_state: Optional[robothub_core.DeviceState] = None
+        self.__device: Optional[OakCamera] = None
+        self.__device_thread: Optional[Thread] = None
 
-            device_thread = Thread(target=self.__manage_device,
-                                   kwargs={'device': device},
-                                   name=f'connection_{device_mxid}')
-            device_thread.start()
-            self.__camera_threads.append(device_thread)
+        atexit.register(self.__cleanup)
+        signal.signal(signal.SIGUSR1, self.__cleanup)
```

Comment:
because `atexit` doesnt get triggered by SIGUSR1. I want the same logic as in the controller, atexit handles unexpected app termination and SIGUSR1 handles Stop button pushed by user.

Reply:
But `atexit` gets triggered by normal interpreter termination ([atexit](https://docs.python.org/3/library/atexit.html)), so when app receives SIGUSR1 it exits normally (it's not killed).

Reply:
when app receives SIGUSR1 then the `robothub-core` calls the `_stop` method. This method checks if there are no more threads running, if there are, it exits with error. atexit is called only after the _stop method finishes (when it trigger the `os._exit(exc)` or `exit(exc)` function.

I think the better way of doing this would be to call the cleanup in the `on_stop` method, but then we would not have any `on_stop` method for the users. So we decided to keep the on_stop method for users and we duplicated the robothub-core logic

Reply:
I didn't know `robothub-core` calls `os._exit`. When `os._exit` is called no cleanup functions are executed, I think it's a mistake and we should fix it. And yep, I agree that our cleanup function should be executed before `_stop` checks number of threads.

I think we are a little bit inconsistent with `on_start` and `on_stop` methods. Why did we overwrite the first and didn't want to touch the second? What if user wants to implement their own `on_start`?

Reply:
what would you do instead of calling `os._exit`?

The way I think about it is that basic user will not need to overwrite either method. he will just impleemnt the pipeline creation method and create objects in the init method.

For advanced users who spawn threads/processes - they will need to implement the `on_stop` method so that the app terminates cleanly. I don't see any reason why anyone should overwrite the `on_start` method. That would be probably just a user who wants to work with multiple devices and this user will need to use the `robothub_core` and not `robothub` anyway

Thoughts?

Reply:
I would call `sys.exit` as it's a recommended way to tell the interpreter to exit, as it allows cleanup (`atexit`) functions to be run.

I completely agree with your use case description. It's probably okay to leave it the way it is now.

Reply:
ok